Upgrade webpack and convert to TS

georgyangelov opened the pull request:
@cwillisf I am weak. I went down the rabbit hole (again). I found the rabbit... actually, I found at least 4-5 hostile rabbits there. 😅 

So I didn't want to have ScratchStorage as `any` so I decided to add the types here. But scratch-storage uses webpack 4 so I thought "how hard can it be to upgrade it". Which lead to web workers packaging which changed in 5 (it's a separate chunk now, not inlined into the main bundle). Which is a breaking change for the lib so => new major version.

Had to also fight jest for the workers, and some specifics around how TS generates `module.exports.default = ...` and not `module.exports = ...` so I changed some exports to be non-default (don't really like default ones https://blog.gangelov.net/posts/why-named-exports/ ).

Anyway, I have this PR with what I believe are no (or minimal) behavior changes, but we have typings which can be used by `scratch-web` for the NGP.

I planned to change a bit the saving logic so that it's not directly going to the webHelper, but allows for configurability, so that we can have, for example, the IndexedDB store I hacked around locally (don't ask - `storage.webHelper = { ... }` 😅).
Anyway - I'll see if this is necessary for now. And that would be a separate PR. I'm thinking maybe something similar to express middlewares - where each store can either proceed to the next one or short-circuit. Does that sound alright on a higher-level?

Also I realize I haven't asked you whether you're ok with having TS in here (or the other changes), so if you have any concerns - I'm definitely open to a discussion.

georgyangelov committed:
feat!: upgrade webpack to 5 and add TS support

Breaking flag is because it may have some differences in the way the library is exported -
`module.exports = ` vs `module.exports.default = `. That would depend on the Webpack config,
so it should continue working, but just to be safe.

georgyangelov committed:
chore: name the worker chunk

georgyangelov committed:
chore: add types

georgyangelov committed:
chore: make it lint TS

georgyangelov committed:
chore: remove leftover no-cache

georgyangelov committed:
chore: remove leftover comments

georgyangelov committed:
chore: fix the asset id types -> seems like they can be numbers?

georgyangelov committed:
chore: export some types

georgyangelov commented:
Okay, decided that it's not worth doing any more changes now. With that this set of changes only stays as "upgrade webpack and add typings", so @cwillisf if you prefer to give this more time to mature, I can add some temporary types in `scratch-gui` or `scratch-web` directly and not have to rely on this PR at all. Would that be better?

See https://github.com/scratchfoundation/scratch-editor/pull/32 and https://github.com/scratchfoundation/scratch-platform/pull/60 for how it's used in the current state.

cwillisf commented on jest.config.js:
```diff
@@ -1,5 +1,37 @@
+            // `new Worker(new URL('./worker.js', import.meta.url));`.
+            // See https://webpack.js.org/guides/web-workers/
+            // However, the `import.meta.url` is ESM-only and Jest's support for ESM is
+            // still experimental. So, we need to mock it instead (or use experimental
+            // jest & node features).
```
File this under "justifications for future migration to Playwright"?

georgyangelov replied:
Probably 🤷  Depends on what level we're testing on - E2E through the UI - definitely. Unit-level JS logic - jest still seems the best option

cwillisf commented on package.json:
```diff
@@ -11,8 +11,9 @@
+  "types": "./dist/types/index.d.ts",
   "scripts": {
-    "build": "webpack --progress --colors --bail",
+    "build": "rm -rf dist && webpack",
```
```suggestion
    "build": "rimraf dist && webpack",
```
for Windows compatibility

georgyangelov replied:
Ah, I knew there had to be some reason to use an npm module 😄

cwillisf commented on src/Asset.ts:
```diff
@@ -0,0 +1,86 @@
+import {Buffer} from 'buffer';
+import {DataFormat} from './DataFormat';
+
+// TODO: The comments in this file indicate that the asset id is a string only, but
+// the types in BuiltinHelper and the default project builder in scratch-gui
+// allow for it to be a number as well.
```
This one's a good advertisement for TS -- or at least good JSDoc hints.

I think `string` is the "correct" option (for now: a string that looks like an integer; later: probably a string that looks like a UUID). I think this is a reasonable approach given our current reality. Thanks for identifying where the numbers are coming from!

georgyangelov replied:
I believe the only **value** that I saw setting it to a number is this in defaultProject in GUI:

```
{
        id: 0,
        assetType: 'Project',
        dataFormat: 'JSON',
        data: JSON.stringify(projectJson)
    }
```

Should I change that to `"0"` and make the AssetId type be just `string`?

cwillisf commented on src/DataFormat.ts:
```diff
@@ -11,6 +11,6 @@ const DataFormat = {
+} as const;
 
-module.exports = DataFormat;
+export type DataFormat = typeof DataFormat[keyof typeof DataFormat];
```
What does `typeof DataFormat[keyof typeof DataFormat]` do? Is that a common TS idiom?

georgyangelov replied:
Ah, yes, it's basically saying "give me the type of all possible values that DataFormat can have". So that would be the type `'"jpg" | "json" | "mp3" | "png" | "sb2" | "sb3" | "svg" | "wav"`. We can also say `string`, but since they are an enumeration I wanted to be more specific just in case somewhere this pays off.

In  `typeof DataFormat[keyof typeof DataFormat]`:
- `typeof DataFormat` is `{JPG: 'jpg', JSON: 'json', ...}`
- `keyof {JPG: 'jpg', JSON: 'json', ...}` is `"JPG" | "JSON" | ...` - basically the type of all possible keys as a union
- `{JPG: 'jpg', JSON: 'json'}['JPG']` is `'jpg'` - like getting the value of a property, but for types
- And you can have unions as the indexing: `{JPG: 'jpg', JSON: 'json', PNG: 'png'}['JPG' | 'PNG']` gets you `'jpg' | 'png'` (basically following all versions of the union)

cwillisf commented on src/FetchWorkerTool.ts:
```diff
@@ -33,22 +45,26 @@ class PrivateFetchWorkerTool {
                     }
                     for (const message of data) {
-                        if (this.jobs[message.id]) {
+                        const job = this.jobs[message.id];
```
😎

georgyangelov replied:
Had to please TS here since it's an access and it doesn't know that it's non-null later 😄

cwillisf commented on src/ScratchStorage.ts:
```diff
@@ -182,15 +200,14 @@ class ScratchStorage {
+    load (assetType: AssetType, assetId: AssetId, dataFormat: DataFormat): Promise<_Asset | null> {
         const helpers = this._helpers.map(x => x.helper);
-        const errors = [];
+        const errors: unknown[] = [];
```
I'm still a little embarrassed that I didn't use `Error` throughout :/

georgyangelov replied:
I didn't check, to be honest, but having promise rejections be `unknown` in TS is standard practice since any rejection up the chain can be an arbitrary object and this can even be libraries. So even if we use Error all around our code, it still is hard to make sure it's an Error in every possible case.

cwillisf commented on src/memoizedToString.js:
```diff
@@ -0,0 +1,75 @@
+    _TextEncoder = TextEncoder;
+}
+
+const memoizedToString = (function () {
```
Thanks for pulling this out of `Asset`!

georgyangelov replied:
I didn't want to have the `require` above in TS and thought it safer to leave in JS for now :D

cwillisf reviewed (approved):
Awesome!

"I am weak." :laughing:
I guess your weakness leads to strength!

Thanks for doing this. It looks great! My only request is to change `rm -rf` to `rimraf` (and add that as a dev dep) to help developers on Windows.

cwillisf commented:
> I planned to change a bit the saving logic so that it's not directly going to the webHelper, but allows for configurability

> I'm thinking maybe something similar to express middlewares - where each store can either proceed to the next one or short-circuit. Does that sound alright on a higher-level?

That all sounds good to me! The Express-style middleware pattern seems like a good choice, both for the way it works and for familiarity among JS / TS devs.

georgyangelov committed:
chore: use rimraf & make webpack build in prod mode always

georgyangelov commented:
Thank you for reviewing and being open to bigger changes! As a next step, I'll finish up the scratch-gui changes and make 100% sure all tests in there pass as well.

Do you think I should test in some other way before merging this? Some other project that uses scratch-storage?
Or something to document somewhere (readme doesn't seem to need changes, but maybe document something?)

georgyangelov committed:
chore: correct more types, turns out `data` should be `Uint8Array` instead of `Buffer`

The Buffer is a subclass of Uint8Array, so that's fine to change everywhere, even if
in reallity Buffer is used for some of the assets.

georgyangelov committed:
chore: asset data can be a string as well

georgyangelov committed:
chore!: can't have a default and non-default exports in commonjs

BREAKING CHANGE: The ScratchStorage constructor is now exported
as a named export instead of a default export.

georgyangelov committed:
Merge remote-tracking branch 'upstream/develop' into nextgen-support

georgyangelov commented:
I think I'm done with this, I corrected a couple more types and fixed the node bundle that I broke previously. https://github.com/scratchfoundation/scratch-storage/pull/649/commits/ff5b428e7152b97671262da1d7ac08f3abfe7259 this is a breaking change as well, but I guess it should be fine to not update the non-monorepo scratch-gui?

I can also backport the change, what do you think?

georgyangelov committed:
Merge remote-tracking branch 'upstream/develop' into nextgen-support
